store/sqlstore/migration.go

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sqlstore
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+const (
+	MigrationDefaultLockTimeout = 3
+	MigrationDefaultBackoffTime = 5
+	MigrationDefaultNumRetries  = 3
+)
+
+type AsyncMigrationStatus string
+
+const (
+	AsyncMigrationStatusUnknown  AsyncMigrationStatus = ""
+	AsyncMigrationStatusRun      AsyncMigrationStatus = "run"      // migration should be run
+	AsyncMigrationStatusSkip     AsyncMigrationStatus = "skip"     // migration should be skipped (not sure if needed?)
+	AsyncMigrationStatusComplete AsyncMigrationStatus = "complete" // migration was already executed
+	AsyncMigrationStatusFailed   AsyncMigrationStatus = "failed"   // migration has failed
+)
+
+// AsyncMigration executes a single database migration that allows concurrent DML
+type AsyncMigration interface {
+	// name of migration, must be unique among migrations - used for saving status in database
+	Name() string
+	// returns if migration should be run / was already executed
+	GetStatus(SqlStore) (AsyncMigrationStatus, error)
+	// exectutes migration, gets started transaction with lock timeouts set
+	Execute(context.Context, SqlStore, *sql.Conn) (AsyncMigrationStatus, error)
+}
+
+// MigrationRunner runs queued async migrations
+type MigrationRunner struct {
+	supplier   SqlStore
+	migrations []AsyncMigration
+	done       chan struct{}
+	options    MigrationOptions
+	cancelFn   context.CancelFunc
+}
+
+type MigrationOptions struct {
+	LockTimeout int // Lock timeout in seconds
+	BackoffTime int // Time to wait between retries
+	NumRetries  int // Number of retries in case of failed migration
+}
+
+func NewMigrationRunner(s SqlStore, o MigrationOptions) *MigrationRunner {
+	if o.LockTimeout == 0 {
+		o.LockTimeout = MigrationDefaultLockTimeout
+	}
+	if o.BackoffTime == 0 {
+		o.BackoffTime = MigrationDefaultBackoffTime
+	}
+	if o.NumRetries == 0 {
+		o.NumRetries = MigrationDefaultNumRetries
+	}
+	return &MigrationRunner{
+		supplier: s,
+		done:     make(chan struct{}),
+		options:  o,
+	}
+}
+
+// Add checks if the migration should be executed and adds it to queue
+func (r *MigrationRunner) Add(m AsyncMigration) error {
+	// check status in Systems table
+	currentStatus, appErr := r.supplier.System().GetByName("migration_" + m.Name())
```

Comment:
I guess the need to check status arises from the fact that we are passing migration structs which can themselves be executed before being added to the queue ?

In that case, I would suggest creating a factory function which passes the initialization code in Add, which is then actually executed inside `Run` and then the `m.execute` is called. This ensures that migrations are instantiated only inside the MigrationRunner and nowhere else. Or atleast, you can't pass an instantiated migration into the runner.

Take a look at https://github.com/mattermost/mattermost-load-test-ng/blob/master/cmd/loadtest/loadtest.go#L22 if you want to get an idea of what I'm hinting of. Let me know if I wasn't able to explain.

Reply:
This allows for example to manually disable particular migration. I wouldn't be worried by execution of migrations outside of runner.  Factory function would make it a bit more complicated,  but I will think about that.